contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol

```diff
@@ -0,0 +1,843 @@
+        uint256 currentPoolWethShare,
+        uint256 targetPoolWethShare,
+        uint256 WETHPositionBalance,
+        uint256 OETHbPositionBalance
+    );
+    
+    event PoolWethShareUpdated(
+        uint256 newWethShare
+    );
+
+    event PrincipalPositionBeforeSwap(
+        uint256 WETHPositionBalance,
+        uint256 OETHbPositionBalance
+    );
+
+    event PrincipalPositionAfterSwap(
+        uint256 WETHPositionBalance,
+        uint256 OETHbPositionBalance
+    );
+
+    event WithdrawLiqiudityShareUpdated(
+        uint128 newWithdrawLiquidityShare
+    );
+
+    event PoolWethShareVarianceAllowedUpdated(
+        uint256 poolWethShareVarianceAllowed
+    );
+
+    event LiquidityRemoved(
+        uint128 withdrawLiquidityShare,
+        uint256 removedWETHAmount,
+        uint256 removedOETHbAmount,
+        uint256 WETHAmountCollected,
+        uint256 OETHbAmountCollected,
+        uint256 netValue
+    );
+
+    event LiquidityTokenBurned(
+        uint256 tokenId
+    );
+
+    event LiquidityAdded(
+        uint256 WETHAmountDesired,
+        uint256 OETHbAmountDesired,
+        uint256 WETHAmountSupplied,
+        uint256 OETHbAmountSupplied,
+        uint256 tokenId,
+        uint256 netValue
+    );
+
+    /**
+     * @dev Verifies that the caller is the Governor, or Strategist.
+     */
+    modifier onlyGovernorOrStrategist() {
+        require(
+            msg.sender == governor() ||
+            msg.sender == IVault(vaultAddress).strategistAddr(),
+            "Not the Governor or Strategist"
+        );
+        _;
+    }
+
+    /// @notice the constructor
+    /// @param _stratConfig the basic strategy configuration
+    /// @param _wethAddress Address of the Erc20 WETH Token contract
+    /// @param _OETHbAddress Address of the Erc20 OETHb Token contract
+    /// @param _swapRouter Address of the Aerodrome Universal Swap Router
+    /// @param _nonfungiblePositionManager Address of position manager to add/remove
+    ///         the liquidity
+    /// @param _clPool Address of the Aerodrome concentrated liquidity pool
+    /// @param _sugarHelper Address of the Aerodrome Sugar helper contract
+    constructor(
+        BaseStrategyConfig memory _stratConfig,
+        address _wethAddress,
+        address _OETHbAddress,
+        address _swapRouter,
+        address _nonfungiblePositionManager,
+        address _clPool,
+        address _sugarHelper
+    ) InitializableAbstractStrategy(_stratConfig)
+    {
+        WETH = _wethAddress;
+        OETHb = _OETHbAddress;
+        swapRouter = ISwapRouter(_swapRouter);
+        positionManager = INonfungiblePositionManager(
+            _nonfungiblePositionManager
+        );
+        clPool = ICLPool(_clPool);
+        helper = ISugarHelper(_sugarHelper);
+        sqrtRatioX96Tick0 = ISugarHelper(_sugarHelper).getSqrtRatioAtTick(-1);
+        sqrtRatioX96Tick1 = ISugarHelper(_sugarHelper).getSqrtRatioAtTick(0);
+
+        lowerTick = -1;
+        upperTick = 0;
```

Comment:
A lot of stuff in the contract depends which token is token0 and which is token1.

We should either make the code able to handle either direction OR have a require that reverts if the tokens are in the wrong order for the code, and require a code upgrade later if we need to use that.

Reply:
There might go more work into generalizing this than I'd like right now. Have made some generalisations here:https://github.com/OriginProtocol/origin-dollar/pull/2184/commits/360240755dadaa27fc1353914e8b0855098e31c3 

Though we will need more testing for complete generalisation. Especially when the token0 & token1 are reversed in the underlying pool. And that is decided by comparing their addresses.

Reply:
Even if we don't make it generalized now, we should make it revert in the constructor if things are wrong.

Reply:
great point. Done here: https://github.com/OriginProtocol/origin-dollar/pull/2184/commits/6eaddbe5e0c65a05436e49be0cf6be61b9cb0214